### PR TITLE
cargo: Fix feature resolution

### PR DESCRIPTION
**Pull request by xclaesse**
Introduce a global Cargo interpreter state that keeps track of enabled
    features on each crate.
    
    Before generating AST of a Cargo subproject, it downloads every
    sub-subproject and resolves the set of features enabled on each of them
    recursively. When it later generates AST for one its dependencies, its
    set of features and dependencies is already determined.

This is based on top of https://github.com/mesonbuild/meson/pull/12945

### CODE REVIEW

**Comment by swick** [comment]
Before this MR it was possible to have a wrap or Cargo.lock for one specific crate which then also has a Cargo.lock with its own dependencies. With this MR, that doesn't seem to be the case anymore.

**Comment by xclaesse** [comment]
@swick I'm not sure exactly what you mean. If the main project has a wrap for a cargo dependency, the Cargo.lock from that dep will be used as well. 

The big limitation this PR leaves is when you have multiple "entry point" into cargo. For example when the main project does :
```
adep = dependency('a-rs')
bdep = dependency('b-rs')
```

If `a` and `b` have a common `c-rs` crate, it will be configured with whatever features `a` needs. If `b` then needs extra features that's an error.

The easy workaround for that would be to have a single entry point, for example by adding `rust.dependencies('a-rs', 'b-rs')`.

**Comment by xclaesse** [comment]
> @swick I'm not sure exactly what you mean. If the main project has a wrap for a cargo dependency, the Cargo.lock from that dep will be used as well.

Actually, reading the code again (been a while), it's a bit more complicated than that. The way Cargo works is the top level Cargo.lock takes precedence over all other dependencies. The main project's Cargo.lock is supposed to contain the whole dependency tree. It may not work exactly that way in Meson currently... need to double check.

**Comment by xclaesse** [comment]
@swick I think I understood what you mean and I added a commit to fix your case. Can you test it again please?

**Comment by swick** [comment]
Same error locally as in CI.

**Comment by swick** [comment]
With that fixed it's behaving as before.

If I have tow dependencies, their features are not unified though, are they? This is what I currently get:

```
subprojects/serde_derive-1.0.102/meson.build:67:-1: ERROR: Problem encountered: Dependency syn-1-rs previously configured with features ['quote', 'default', 'clone-impls', 'derive', 'parsing', 'printing', 'proc-macro'] but need ['quote', 'visit', 'default', 'clone-impls', 'derive', 'parsing', 'printing', 'proc-macro']
```
I'm also failing to turn on the specific feature:
```
project(...
  default_options: {
    'syn-1-rs:feature-visit': true,
  },
)
```
```
subprojects/syn-1.0.82/meson.build:1:-1: ERROR: In subproject syn-1-rs: Unknown options: "syn-1-rs:feature-visit"
```

**Comment by xclaesse** [comment]
> If I have tow dependencies, their features are not unified though, are they?

They are not unified, that's the limitation I mentioned that could be worked around by adding e.g. `rust.dependencies('a-rs', 'b-rs')`.

> I'm also failing to turn on the specific feature:

This PR removes those per subproject feature options because features are global. I think we should add a global module option like `-Drust.features=foo,bar`. More similar to cargo: https://doc.rust-lang.org/cargo/reference/features.html#command-line-feature-options.

As workaround for now, you can create a dummy cargo project that depends on both crates you need, and use that as meson subproject.

**Comment by swick** [comment]
> As workaround for now, you can create a dummy cargo project that depends on both crates you need, and use that as meson subproject.

Yeah, that seems to work. Now the biggest issue seems to be `target.cfg.dependencies` not working.

File: docs/markdown/snippets/cargo_features.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+This solves many cases of Cargo subprojects being configured with missing
+features that the main project had to enable by hand using e.g.
+`default_options: ['foo-rs:feature-default=true']`.
+
+Note that there could still be issues in the case there are multiple Cargo
+entry points. That happens if the main Meson project makes multiple `dependency()`
+calls for different Cargo crates that have common dependencies.
+
+Breaks: This change removes per feature Meson option That were previously
+possible to set as show above or from command line `-Dfoo-rs:feature-foo=true`.
```
Comment (eli-schwartz) [inline docs/markdown/snippets/cargo_features.md:14]: Wouldn't this also break projects that are setting those options via `default_options: [...]`?

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: docs/markdown/snippets/cargo_features.md (hunk 1 above)
Comment (xclaesse) [reply docs/markdown/snippets/cargo_features.md:14]: Yes that's an incompatible change for cargo subprojects, but we never claimed that was stable. Hopefully not many projects actually uses it yet.

Comment (eli-schwartz) [reply docs/markdown/snippets/cargo_features.md:14]: I'm looking and I cannot find a single place where we claim it is *not* stable, either as a meson warning (the way we print for `import()` of an unstable module) or in the documentation.

I think the technical term for this is "we blew it".

Comment (xclaesse) [reply docs/markdown/snippets/cargo_features.md:14]: It's here: https://github.com/mesonbuild/meson/pull/12358. I asked many times to merge that before release...

Comment (xclaesse) [reply docs/markdown/snippets/cargo_features.md:14]: > I think the technical term for this is "we blew it".

yes... :/

Comment (xclaesse) [reply docs/markdown/snippets/cargo_features.md:14]: Actually my PR for that was https://github.com/mesonbuild/meson/pull/12356 but then @dcbaker wanted to make a more complex but better solution, and we lost track if it... :/

Comment (dcbaker) [reply docs/markdown/snippets/cargo_features.md:14]: Yeah, we did. I meant to just merge 12356 and replace it later if I ever got 12358 working. That's on me

File: mesonbuild/cargo/interpreter.py [hunk 2]
```diff
@@ -556,188 +591,134 @@ def _create_dependencies(cargo: Manifest, build: builder.Builder) -> T.List[mpar
                 ),
                 _dependency_varname(dep.package),
             ),
-            # if xxx_dep.found()
-            build.if_(build.method('found', build.identifier(_dependency_varname(dep.package))), build.block([
-                # actual_features = xxx_dep.get_variable('features', default_value : '').split(',')
-                build.assign(
+            # actual_features = xxx_dep.get_variable('features', default_value : '').split(',')
```
Comment (xclaesse) [inline mesonbuild/cargo/interpreter.py:594]: gfds

**Pull request closed without merging**